Automattic/wp-calypso
# format-currency: Change formatCurrency to use locale

#### Proposed Changes

In calypso when we need to format prices for display, we often use the `@automattic/format-currency` library. The library contains a hard-coded list of currencies and the data needed to format them appropriately for display. It exports a `formatCurrency()` function to perform this formatting. However, the data this system uses is entirely based on the currency code, and does not take into account the user's locale. Many currencies are formatted differently depending on the locale, which makes this system inaccurate.

For example, sometimes EUR 1000 is formatted `€1,000.00` but other times it is formatted `1.000,00 €`.

In this PR we modify `formatCurrency()` and its sibling `getCurrencyObject()` to ignore the hard-coded list of data and instead use the browser's built-in library that formats currency for locales: [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat). It retrieves the locale from the browser (although it can also be passed-in as an option if needed). 

Before this PR. The variant picker prices are formatted by `formatCurrency` and the line item price is formatted by the backend. Note that in the IDR example the thousands group separators of the variant picker do not match those of the line item.

<img width="569" alt="before-en_US-IDR" src="https://user-images.githubusercontent.com/2036909/209868111-8d2b864a-23b6-4519-a02f-5a20d2afcf24.png">

<img width="568" alt="before-de_DE-EUR" src="https://user-images.githubusercontent.com/2036909/209868617-6c9c195e-a3e1-4c61-a149-2ceb76a151b4.png">

After this PR. Note that the thousands separators now are correct in both the frontend and backend versions of the price.

<img width="572" alt="Screenshot 2023-01-05 at 1 11 51 PM" src="https://user-images.githubusercontent.com/2036909/210850841-61c035f3-32d5-410c-917d-d4f47e593689.png">

<img width="571" alt="Screenshot 2022-12-28 at 5 09 34 PM" src="https://user-images.githubusercontent.com/2036909/209877326-86326854-8c92-4922-92fd-42d124746edb.png">


**NOTE:** a possible improvement would be to pass the locale from WordPress to the library instead of relying on the browser. This PR adds a `setDefaultLocale()` function which could be used to consume the `getCurrentLocaleSlug()` Redux selector, but I'm not sure where that side effect would best be placed.

**NOTE:** some of the official currency symbols returned by using `Intl.NumberFormat` might be confusing for users in their own locale. For example, if a user has their locale set to `en_CA` but their currency set to `CAD`, they would see prices displayed like `$100` instead of `C$100` as they do before this PR and that could be misinterpreted as USD (although if their locale is `en_US` they would see `CA$100`). To make this change less disruptive, the PR retains the hard-coded list of currency _symbols_ and uses those as overrides for each currency, _regardless of locale_ (except for USD because of https://github.com/Automattic/wp-calypso/issues/46205).

This PR removes the hard-coded formatting data from the package, and the unnecessary `CURRENCIES` and `getCurrencyDefaults()` exports. It also removes the `decimal`, `grouping`, and `symbol` options from `formatCurrency()` and `getCurrencyObject()` because those are now totally under the control of the formatter.

Part of https://github.com/Automattic/payments-shilling/issues/1300

#### Testing Instructions

Automated tests should mostly cover this, but it's worth viewing a few significant places in calypso that display currencies to make sure there are no obvious problems. Please do this with different currencies and different locales.

- The plans page.
- Checkout.

As for the changes in the API, you can verify that there are no uses of the removed options or exports with the following commands. Make sure that you check any results to be sure they are not using the removed options or exports:

- `rg --context 10 -w formatCurrency |rg -w 'grouping|decimal|symbol'`
- `rg --context 10 -w getCurrencyObject |rg -w 'grouping|decimal|symbol'`
- `rg -w CURRENCIES`
- `rg -w getCurrencyDefaults`
